_posts/2021-03-23-activesupport.html

```diff
@@ -0,0 +1,172 @@
+      </li>
+      <li>
+        <a href="/blog/influxdb-rails.html">Measure twice, instrument once with influxdb-rails!</a>
+      </li>
+      <li>
+        <a href="/blog/beyond-performance.html">influxdb-rails instrumentation beyond performance.</a>
+      </li>
+    </ul>
+  </p>
+</div>
+
+<h2>Publish &amp; Subscribe</h2>
+
+<p>
+  There is an instrumentation message emitted from <code>ActionController</code> that includes interesting data
+  about the things happening in your controller action. Let's explore this.
+</p>
+<p>
+  If you don't have a Ruby on Rails app at hand, just setup a minimal one with
+</p>
+<p>  
+  <code>rails new --minimal</code>
+</p>
+
+<p>
+  Add this code into an initializer
+</p>
+
+{% highlight ruby %}
+# config/initializers/instrumentation_subscriber.rb
+
+# What happens in ActionController?
+module ActionController
+  class InstrumentationSubscriber < ActiveSupport::Subscriber
+    attach_to :action_controller
+
+    def process_action(message)
+      Rails.logger.debug "Instrumentation #{message.name} Received!"
+    end
+  end
+end
+{% endhighlight %}
+
+<p>
+  Boot the app (<code>rails server</code>), visit <a href="http://127.0.0.0:3000">http://127.0.0.0:3000</a> and you'll
+  see the new log lines. So what? What's the difference to calling <code>Rails.logger</code> in an action or in a callback
+  inside your controller? Why is <code>ActiveSupport::Notifications</code> fabulous?
+</p>
+
+<h2>ActiveSupport::Notifications Scales</h2>
+<p>
+  First, as explained in the intro, the main advantage is that the publisher is decoupled from the subscriber. For instance,
+  you can have more than one subscriber listening to <code>process_action.action_controller</code>.
+</p>
+
+{% highlight ruby %}
+# config/initializers/slowlog_subscriber.rb
+
+module ActionController
+  class SlowlogSubscriber < ActiveSupport::Subscriber
+    attach_to :action_controller
+
+    def process_action(message)
+      return if message.duration <= 10
+
+      controller_location = [message.payload[:controller], message.payload[:action]].join("#") # -> ThingsController#something
+      Rails.logger.debug "#{controller_location} was slow (#{message.duration}ms)" 
+    end
+  end
+end
+{% endhighlight %}
+
+<p>
+  You are free to organize this however you want. Decouple publisher/subscriber in different files, chronological or
+  even in different threads.
+</p>
+
+{% highlight ruby %}
+# config/initializers/poor_mans_background_job_subscriber.rb
+
+module ActionController
+  class PMBGJSubscriber < ActiveSupport::Subscriber
+    include Concurrent::Async
+
+    attach_to :action_controller
+
+    def process_action(message)
+      async.background_job(message)
+    end
+
+    def background_job(message)
+      # ...do something expensive with the message in a thread
+      sleep(60)
+    end
+  end
+end
+{% endhighlight %}
+
+<h2>ActiveSupport::Notifications is everywhere today</h2>
```

Comment:
Another capitalization issue. Check other titles too.

```suggestion
<h2>Activesupport::Notifications Is Everywhere Today</h2>
```